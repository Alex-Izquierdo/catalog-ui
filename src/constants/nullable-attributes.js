export const PORTFOLIO_ITEM_NULLABLE = [
  'display_name', 'description', 'long_description', 'distributor', 'documentation_url', 'support_url', 'workflow_ref'
];
