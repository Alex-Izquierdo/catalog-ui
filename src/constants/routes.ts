export const ADD_PORTFOLIO_ROUTE = '/portfolios/add-portfolio';
export const ADD_PRODUCTS_ROUTE = '/portfolios/add-products';
export const EDIT_ORDER_PROCESS_ROUTE = '/portfolios/edit-order-process';
export const EDIT_PORTFOLIO_ROUTE = '/portfolios/edit';
export const EDIT_PORTFOLIO_WORKFLOW_ROUTE = '/portfolios/edit-workflow';
export const EDIT_SURVEY_ROUTE = '/portfolios/portfolio-item/edit-survey';
export const NESTED_EDIT_ORDER_PROCESS_ROUTE =
  '/portfolios/portfolio/edit-order-process';
export const NESTED_EDIT_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/edit-portfolio';
export const NESTED_REMOVE_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/remove-portfolio';
export const NESTED_SHARE_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/share-portfolio';
export const NESTED_WORKFLOW_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/edit-workflow';
export const ORDER_ROUTE = '/orders/order';
export const ORDERS_ROUTE = '/orders';
export const PLATFORM_INVENTORIES_ROUTE = '/platforms/platform-inventories';
export const PLATFORM_ROUTE = '/platforms/platform';
export const PLATFORM_SERVICE_OFFERINGS_ROUTE = '/platforms/service-offerings';
export const PLATFORM_TEMPLATES_ROUTE = '/platforms/platform-templates';
export const PLATFORM_DETAILS_ROUTE = '/platforms/platform-details';
export const PLATFORMS_ROUTE = '/platforms';
export const PORTFOLIO_ROUTE = '/portfolios/portfolio';
export const PORTFOLIO_ITEM_ROUTE = '/portfolios/portfolio-item';
export const PORTFOLIO_ITEM_ORDER_ROUTE = '/portfolios/portfolio-item/order';
export const PORTFOLIO_ITEM_ROUTE_EDIT = '/portfolios/portfolio-item/edit';
export const PORTFOLIO_ITEM_EDIT_ORDER_PROCESS_ROUTE =
  '/portfolios/portfolio-item/edit-order-process';
export const PORTFOLIOS_ROUTE = '/portfolios';
export const REMOVE_PORTFOLIO_ROUTE = '/portfolios/remove';
export const SHARE_PORTFOLIO_ROUTE = '/portfolios/share';
export const WORKFLOW_PORTFOLIO_ROUTE = '/portfolios/edit-workflow';
export const ORDER_PROCESSES_ROUTE = '/order-processes';
export const ADD_ORDER_PROCESS_ROUTE = '/order-processes/add-order-process';
export const REMOVE_ORDER_PROCESS_ROUTE = '/order-processes/remove';
export const UPDATE_ORDER_PROCESS_ROUTE = '/order-processes/update';
