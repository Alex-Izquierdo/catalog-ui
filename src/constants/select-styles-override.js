const style = {
  container: provided => ({
    ...provided,
    '&:focus': {
      outline: 'none'
    }
  }),
  control: (provided, { isFocused }) => ({
    ...provided,
    borderRadius: 0,
    borderColor: '#EDEDED',
    backgroundColor: '#FFFFFF',
    borderBottomColor: isFocused ? '#007bba' : '#282d33',
    cursor: 'pointer',
    boxShadow: 'none',
    minHeight: 36,
    width: 256,
    '&:hover': {
      borderBottomColor: '#007bba'
    }
  }),
  indicatorSeparator: () => ({}),
  dropdownIndicator: provided => ({
    ...provided,
    color: '#282d33',
    paddingBottom: 6,
    paddingTop: 6
  }),
  menu: provided => ({
    ...provided,
    borderRadius: 'none',
    boxShadow:
      '0 0.0625rem 0.0625rem 0rem rgba(3,3,3,.05),0 0.25rem 0.5rem 0.25rem rgba(3,3,3,.06)',
    paddingBottom: '0.5rem',
    paddingTop: '0.5rem',
    marginTop: 4,
    width: 'calc(100% - 4px)',
    marginRight: 2,
    marginLeft: 2
  }),
  option: (provided, { isFocused }) => ({
    ...provided,
    cursor: 'pointer',
    textAlign: 'left',
    color: '#282d33',
    fontSize: '1rem',
    fontWeight: '400',
    lineHeight: '1.5',
    paddingBottom: '0.5rem',
    paddingTop: '0.5rem',
    paddingLeft: '1rem',
    paddingRight: '1rem',
    backgroundColor: isFocused ? '#ededed' : 'transparent',
    '&:hover': {
      backgroundColor: '#ededed'
    }
  }),
  placeholder: provided => ({
    ...provided,
    left: 20
  }),
  singleValue: provided => ({
    ...provided,
    left: 20,
    maxWidth: 'calc(100% - 28px)'
  })
};

export default style;
