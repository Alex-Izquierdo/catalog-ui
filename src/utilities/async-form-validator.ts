import awesomeDebouncePromise, {
  AwesomeDebounceOptions
} from 'awesome-debounce-promise';

const asyncFormValidator = (
  asyncFunction: (...args: any[]) => any,
  debounceTime = 250,
  options: Partial<AwesomeDebounceOptions> = { onlyResolvesLast: false }
): any => awesomeDebouncePromise(asyncFunction, debounceTime, options);

export default asyncFormValidator;
