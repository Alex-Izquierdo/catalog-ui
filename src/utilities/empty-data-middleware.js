const emptyDataMiddleware = () => (dispatch) => (action) => {
  const nextAction = { ...action };
  if (
    action.type.match(/_FULFILLED$/) &&
    action.payload &&
    action.meta &&
    action.payload.data &&
    action.payload.meta
  ) {
    nextAction.payload.meta.noData =
      nextAction.payload.meta.count === 0 &&
      nextAction.meta.filter.length === 0;
    return dispatch(nextAction);
  }

  return dispatch(nextAction);
};

export default emptyDataMiddleware;
