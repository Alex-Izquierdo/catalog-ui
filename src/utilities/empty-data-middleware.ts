import { Dispatch } from 'redux';
import { ReduxAction } from '../types/common-types';

const emptyDataMiddleware = () => (dispatch: Dispatch) => (
  action: ReduxAction
): ReduxAction => {
  const nextAction = { ...action };
  if (
    action.type.match(/_FULFILLED$/) &&
    action.payload &&
    action.meta &&
    action.payload.data &&
    action.payload.meta
  ) {
    const noFilter = Object.prototype.hasOwnProperty.call(
      nextAction.meta,
      'filters'
    )
      ? Object.values(nextAction.meta?.filters || {}).every(
          (value?: string) => typeof value === 'undefined' || value.length === 0
        )
      : nextAction.meta?.filter?.length === 0;

    nextAction.payload.meta.noData =
      nextAction.payload.meta.count === 0 && noFilter;
    return dispatch(nextAction);
  }

  return dispatch(nextAction);
};

export default emptyDataMiddleware;
