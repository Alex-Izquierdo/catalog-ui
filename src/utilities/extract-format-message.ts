import { I18nFormatMessage } from '../redux/reducers/i18n-reducer';
import { GetReduxState } from '../types/redux';

const extractFormatMessage = (getState: GetReduxState): I18nFormatMessage =>
  getState().i18nReducer.formatMessage!;

export default extractFormatMessage;
