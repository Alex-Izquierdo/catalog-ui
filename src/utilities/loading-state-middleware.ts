import { SET_LOADING_STATE } from '../redux/action-types';
import { Dispatch } from 'redux';
import { ReduxAction } from '../types/common-types';

const loadingStateMiddleware = () => (dispatch: Dispatch) => (
  action: ReduxAction
): ReduxAction => {
  if (action.type.match(/_REJECTED$/)) {
    dispatch({ type: SET_LOADING_STATE, payload: false });
  }

  return dispatch(action);
};

export default loadingStateMiddleware;
