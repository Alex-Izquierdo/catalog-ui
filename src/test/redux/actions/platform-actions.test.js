import configureStore from 'redux-mock-store' ;
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware, ADD_NOTIFICATION } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import {
  FETCH_PLATFORM,
  FETCH_PLATFORMS,
  FETCH_PLATFORM_ITEMS,
  FETCH_MULTIPLE_PLATFORM_ITEMS
} from '../../../redux/action-types';
import {
  fetchPlatforms,
  fetchPlatformItems,
  fetchMultiplePlatformItems,
  fetchSelectedPlatform
} from '../../../redux/actions/platform-actions';
import { TOPOLOGICAL_INVENTORY_API_BASE } from '../../../utilities/constants';

describe('Platform actions', () => {
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;

  beforeEach(() => {
    mockStore = configureStore(middlewares);
  });

  afterEach(() => {
    fetchMock.reset();
  });

  it('should dispatch correct actions after fetching platforms', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });
    apiClientMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources`, mockOnce({
      body: { data: [{
        id: '1',
        name: 'Source 1'
      }]}
    }));

    const expectedActions = [{
      type: `${FETCH_PLATFORMS}_PENDING`
    }, {
      type: `${FETCH_PLATFORMS}_FULFILLED`,
      payload: [ expect.objectContaining({ id: '1', name: 'Source 1' }) ]
    }];

    return store.dispatch(fetchPlatforms()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after failed fetching platforms', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });
    apiClientMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources`, mockOnce({
      status: 500
    }));

    const expectedActions = [
      { type: `${FETCH_PLATFORMS}_PENDING` }, {
        type: `${ADD_NOTIFICATION}`,
        payload: expect.objectContaining({ variant: 'danger' })
      },
      expect.objectContaining({ type: `${FETCH_PLATFORMS}_REJECTED` })
    ];

    return store.dispatch(fetchPlatforms()).catch(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after fetching platform items', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });

    fetchMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1/service_offerings?archived_at=`, { data: [{
      id: '1',
      name: 'Offering 1'
    }]});

    const expectedActions = [{
      type: `${FETCH_PLATFORM_ITEMS}_PENDING`,
      meta: { platformId: '1' }
    }, {
      meta: { platformId: '1' },
      type: `${FETCH_PLATFORM_ITEMS}_FULFILLED`,
      payload: { data: [{ id: '1', name: 'Offering 1' }]}
    }];

    return store.dispatch(fetchPlatformItems('1')).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after fetching platform items fails', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });

    fetchMock.getOnce(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1/service_offerings?archived_at=`, 500);

    const expectedActions = [
      {
        type: `${FETCH_PLATFORM_ITEMS}_PENDING`,
        meta: { platformId: '1' }
      }, {
        type: `${ADD_NOTIFICATION}`,
        payload: expect.objectContaining({ variant: 'danger' })
      },
      expect.objectContaining({ type: `${FETCH_PLATFORM_ITEMS}_REJECTED` })
    ];

    return store.dispatch(fetchPlatformItems('1')).catch(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after fetching multiple platforms items', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });
    fetchMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1/service_offerings?archived_at=`, { data: [{
      id: '1',
      name: 'Offering 1'
    }]});
    fetchMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/2/service_offerings?archived_at=`, { data: [{
      id: '2',
      name: 'Offering 2'
    }]});
    fetchMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/3/service_offerings?archived_at=`, { data: [{
      id: '3',
      name: 'Offering 3'
    }]});

    const expectedActions = [{
      type: `${FETCH_MULTIPLE_PLATFORM_ITEMS}_PENDING`
    }, {
      type: `${FETCH_MULTIPLE_PLATFORM_ITEMS}_FULFILLED`,
      payload: {
        1: { data: [{ id: '1', name: 'Offering 1' }]},
        2: { data: [{ id: '2', name: 'Offering 2' }]},
        3: { data: [{ id: '3', name: 'Offering 3' }]}
      }
    }];

    return store.dispatch(fetchMultiplePlatformItems([ '1', '2', '3' ])).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after fetching multiple platforms items fails', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });
    fetchMock.getOnce(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1/service_offerings?archived_at=`, { data: [{
      id: '1',
      name: 'Offering 1'
    }]});
    fetchMock.getOnce(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/2/service_offerings?archived_at=`, { data: [{
      id: '2',
      name: 'Offering 2'
    }]});
    fetchMock.getOnce(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/3/service_offerings?archived_at=`, 500);

    const expectedActions = [{
      type: `${FETCH_MULTIPLE_PLATFORM_ITEMS}_PENDING`
    }, {
      type: `${ADD_NOTIFICATION}`,
      payload: expect.objectContaining({ variant: 'danger' })
    },
    expect.objectContaining({ type: `${FETCH_MULTIPLE_PLATFORM_ITEMS}_REJECTED` })
    ];

    return store.dispatch(fetchMultiplePlatformItems([ '1', '2', '3' ])).catch(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after fetching selected platform', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });
    apiClientMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1`, mockOnce({
      body: {
        id: '1',
        name: 'Source 1'
      }
    }));

    const expectedActions = [{
      type: `${FETCH_PLATFORM}_PENDING`
    }, {
      type: `${FETCH_PLATFORM}_FULFILLED`,
      payload: expect.objectContaining({ id: '1', name: 'Source 1' })
    }];

    return store.dispatch(fetchSelectedPlatform('1')).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should dispatch correct actions after fetching selected platform fails', () => {
    const store = mockStore({
      platformReducer: {
        isPlatformDataLoading: false
      }
    });
    apiClientMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1`, mockOnce({
      status: 500
    }));

    const expectedActions = [{
      type: `${FETCH_PLATFORM}_PENDING`
    }, {
      type: `${ADD_NOTIFICATION}`,
      payload: expect.objectContaining({ variant: 'danger' })
    },
    expect.objectContaining({ type: `${FETCH_PLATFORM}_REJECTED` })
    ];

    return store.dispatch(fetchSelectedPlatform('1')).catch(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
