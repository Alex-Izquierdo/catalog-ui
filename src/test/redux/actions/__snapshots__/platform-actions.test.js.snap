// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Platform actions should not dispatch any actions if fetching platforms is in progress 1`] = `"Cannot read property 'then' of undefined"`;
